Email reminder tool

widgetOne opened the pull request:
There are some todos and rough edges to this, but I don't want to do more work until we have tried it in operation a bit and seen how we feel about it.

A quick summary of the big idea of this tool is that you define in a yaml file how often people should be updating different types of issues (label = 'emergency' -> daily, label = 'soonish' -> weekly) and then set this in a daily cron someplace and it will pester people to regularly update their github issues. The basic execution looks like this:

/python   ./issue_update_prompter.py [token] spider rentjungle --send

where --send has it send out the emails and the default is to just show which issues need updated. This will get transcribed into a readme if we decide to keep going with this.

Commit:
Initial Commit of Email Reminder Tool

Last we were discussing how to simplify our check-in meetings, and
we had the idea of just having a script prompt use to update issue
offline, rather than doing it at our checkin meeting.

This is the initial attempt at that. It take a config file that
defines intervals at which we want people to be prodded, and then
sends them, and potentially a pm or manager, reminders if an issue has
been sitting for a while. It user interface is similar to the report
tool, where is takes the github token, target repo and organization
as arguements.

There are chunks of this that should be refactored into smaller
functions, and I want to add usage of bunches. However, I have
already done too much for one commit.

Commit:
Refactor for code reuse and adding of assignees

This commit refactored the decorated issue generation process into
two steps: filtration of marked issues and conversion to bunches.
This was important as the bunching/decoration process is needed
by report.py and the email notification functionality, but the
fitration logic is different.

I also changed the assignee column in the shared table preperation
logic to show assignees instead of the assignee. When there was
multiple assignees, the listed assignee for an issue was sometimes
the person who had already passed it on. Granted, the best practice
is to update these as soon as an issues responsibility changes
but for the purposes of automated report I thought it was better
to list all the people that are currently assigned to something.

Commit:
adding the current config file location to gitignore

Commit:
adding pycharm project files location to gitignore

Commit:
dynamic config loc and bunches

I added a 4th optional arguement to be able to specify the config.yaml
fullpath at run time. This was pretty simple. At some point, we
should add schema validation of the yaml file though, so the
code does not at erratically in response to mis-spellings or
formatting in that file.

I also change the timeframe structure into a bunch to slightly simplify the
syntax at the lower levels.

Commit:
Upgraded arg parser and added default debug mode

As this system could shot out email to a lot of semi-random people,
I decided to change its default to a sort of safe mode. If you run
it with just the required arguements, it will print out the emails
it is thinking of sending. If you add --send, it will send out the
emails. To do this, I upgrade the arg parse from clint to argpars
which has also enabled some type validation.

Commit:
removing some redundant default filepath logic

davidthewatson commented on issue_update_prompter.py:
```diff
@@ -0,0 +1,183 @@
+        assignee_emails = [assignee.email for assignee in issue.assignees if assignee.email]
+        if not assignee_emails:
+            pass
+            # todo: what do we want to do with issues that are not assigned? PM or nothing?
```
I'd mark the _triage_ label so that someone can just do a query on issues with the triage label and actually do the triage.

widgetOne replied:
Sounds like a good short-term fix. In the long run we might want to have an entry for PM email and send these and the missing user email addresses to them.

davidthewatson reviewed (changes requested):
This looks really solid. Such a relief to read code that's immediately clear and concise. 

Could you do just two things: 

* Add an example of the email dummied up so as not to reveal the innocent
* Change that pass exception to mark the triage label

It can be merged quickly then. Thanks!

Commit:
Adding triage label and catch-up to catchup

When there is no recipent for an issue, the label
triage will now be added to it. This works with
our current process. In the long run, this script
may want to support an email address entry that
could get a heads up for these.

Also, I changed catch-up to catchup, as I'm not
sure the former worked with bunches.

Commit:
Adding days_multiplier import and enhancing debugs

I added the days multiplier import to the yaml file. This is to aide
catching-up and getting the email system up to date. In our case, our
final standards are much higher than where we are now, but we don't
want people's first interaction with this tool to be getting 100
notifications. Therefore, I'm adding this days_multiplier import
which I can start out at 10x, and then slowly wind down to 1x / gone
once were are at our target level.

Commit:
First draft of README changes

davidthewatson commented on README.md:
```diff
@@ -71,3 +71,34 @@ When it's formatted:
+
+## issue_update_prompter.py
+
+The issue update prompting tool builds off of the basic issue report. It checks the target github repository, identifies issues which should be updated and emails issue tables to those responsible. The indentification of issues require update is based on a yaml config fill that defines how often issues with certain tags should be updated. The emails are sent to the public email addresses associat ...
```
I think this is supposed to be:

"The identification of issues that require update" and "based on a yaml config that defines how often issues" and "require that an smtp server be running on the host machine". Does that sound right?

davidthewatson reviewed (changes requested):
Mostly LGTM - just fixup the doc clarity issue that I asked about and we're done!

Commit:
Correcting language in new tool description

fixing text based on dave's review feedback.

widgetOne commented:
Dave, look good?

davidthewatson commented:
LGTM!
